include/deal.II/lac/precondition.h

```diff
@@ -585,47 +939,32 @@ class PreconditionJacobi : public PreconditionRelaxation<MatrixType>
  * @endcode
  */
 template <typename MatrixType = SparseMatrix<double>>
-class PreconditionSOR : public PreconditionRelaxation<MatrixType>
+class PreconditionSOR
+  : public PreconditionRelaxation<MatrixType,
+                                  internal::PreconditionRelaxation::
+                                    PreconditionSORPreconditioner<MatrixType>>
```

Comment:
> Provided I didn't miss something, this begs the question why don't you simply implement step() and Tstep() in PreconditionType directly?

Because step is not in the responsibility of the preconditioner: 
https://github.com/dealii/dealii/blob/9a871495b09d955a2351afc435b5dd6653bbac99/include/deal.II/lac/precondition.h#L850-L867

but involves both the linear operator and the preconditioner. Sorry for not be clear about this and spiting this up into two PRs, but I thought it would be better to do the refactoring in the first step and only add additional functionality in a second step.

Reply:
Roger.